### PR TITLE
[feat] Non-compulsory BuilderID for BYOB Builders

### LINKED ISSUE #659: Make builder ID not compulsory for BYOB builders hosted on our repo
See original discussion in https://github.com/slsa-framework/slsa-verifier/issues/656
/cc @enteraga6

### PR DESCRIPTION
**Pull request by enteraga6**
/cc @mihaimaruseac 

Based off the prefix of the BuilderID within the provenance, if the builder use to build the artifact is one of the BYOB builders of slsa-framework/slsa-github-generator repo, the --builderid flag is not need and is handled automatically. This was done to increase access to users since before the automatic pickup of the builder-id would get the delegator. 

Test cases that cover verifyProvenance will need to be complete after the v1.8.0 release of slsa-framework/slsa-github-generator.

closes #659

### CODE REVIEW

File: verifiers/internal/gha/verifier.go [hunk 1]
```diff
@@ -71,15 +71,8 @@ func verifyEnvAndCert(env *dsse.Envelope,
 	// There is a corner-case to handle: if the verified builder ID from the cert
 	// is a delegator builder, the user MUST provide an expected builder ID
 	// and we MUST match it against the content of the provenance.
-	if byob {
-		if builderOpts.ExpectedID == nil || *builderOpts.ExpectedID == "" {
-			// NOTE: we will need to update the logic here once our default trusted builders
```
Comment (laurentsimon) [inline verifiers/internal/gha/verifier.go:76]: why not keep this code and set ExpectedBuilderID if builderOpts.ExpectedID is empty?
```
provenanceOpts.ExpectedBuilderIDPath = "https://github.com/slsa-framework/slsa-github-generator/.github/workflows/"
```

File: verifiers/internal/gha/provenance.go [hunk 2]
```diff
@@ -297,6 +299,45 @@ func VerifyProvenance(env *dsselib.Envelope, provenanceOpts *options.ProvenanceO
 		if err := isValidDelegatorBuilderID(prov); err != nil {
 			return err
 		}
+
+		id, err := prov.BuilderID()
```
Comment (laurentsimon) [inline verifiers/internal/gha/provenance.go:303]: can we simply / replace this new code by:
```
if byob && builderOpts.ExpectedIDPath != nil{
   return verifyBuilderIDPath(prov, provenanceOpts.ExpectedBuilderIDPath) 
}
```

And implement `verifyBuilderIDPath` similarly to `verifyBuilderIDLooseMatch`

File: verifiers/internal/gha/verifier.go (hunk 1 above)
Comment (enteraga6) [reply verifiers/internal/gha/verifier.go:76]: The way the structure of code is, we need provenance to check expected path and verify against `BuilderID` such that we dont get the empty builder error. That is why it got moved into `verifyProvenance`!

File: verifiers/internal/gha/provenance.go (hunk 2 above)
Comment (enteraga6) [reply verifiers/internal/gha/provenance.go]: Yes we can. Added this added this on new branch will close this pr and submit one with this refactor.

**Pull request closed without merging**